aten/src/ATen/native/ForeachUtils.h

```diff
@@ -64,155 +64,85 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
 // - All tensors must be non-overlapping and dense
 // - Resulting tensor must have the same dtype as the input one
 
-// Check if all tensors have the same device, layout, strides and are not overlapping and dense
-bool has_same_attributes(Device expected_device, TensorList tensors) {
-  auto expected_strides = tensors[0].strides();
-  for (const auto& t : tensors) {
-    if (t.device() != expected_device) {
-      return false;
-    }
-
-    if (t.layout() != at::kStrided) {
-      return false;
-    }
-
-    if (!t.is_non_overlapping_and_dense()) {
-      return false;
-    }
-
-    if (t.strides() != expected_strides) {
-      return false;
+bool will_promote_tensor(const Tensor& tensor, Scalar scalar, bool promote_integer_inputs_to_float = false) {
+  // In case of division, integer inputs will result in float
+  if (promote_integer_inputs_to_float) {
+    if (at::isIntegralType(tensor.scalar_type(), /*includeBool*/ true)) {
+      return true;
     }
   }
-
-  return true;
-}
-
-bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
   auto result_dtype = at::result_type(tensor, scalar);
   return result_dtype != tensor.scalar_type();
 }
 
-bool can_use_fast_route(TensorList tensors) {
-#ifdef __HIP_PLATFORM_HCC__
-  return false;
-#else
-  auto expected_device = tensors[0].device();
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+bool check_fast_path_restrictions(
+        ArrayRef<TensorList> tensorLists, 
+        ArrayRef<Scalar> scalarList = {}, 
+        bool promote_integer_inputs_to_float = false) {
```

Comment:
Nit: might be good to document and/or assert preconditions. From what I remember:
- each tensorList in tensorLists must have the same size
- each tensorList in tensorLists must have at least one element
- tensorLists must have at least one element
- scalarList must either be: size 0, size1, or same size as each TensorList in tensorLists

Reply:
Are you talking about

```
// To go via 'fast' path, several conditions must be satisfied	// To go via 'fast' path, several conditions must be satisfied
// - All tensors must be on the same device	// - All tensors must be on the same device
// - All tensors must have strided layout	// - All tensors must have strided layout
// - All tensors must be non-overlapping and dense	// - All tensors must be non-overlapping and dense
// - Resulting tensor must have the same dtype as the input one	// - Resulting tensor must have the same dtype as the input one
```
?
Because that comment doesn't answer my question: what are the pre-conditions that are true before one calls `check_fast_path_restrictions`?

Reply:
Sorry, my bad
These conditions should be satisfied: 

// Check foreach API restrictions 
// - Tensor lists must be non-empty.
// - All tensors in all lists must have the same dtype.
// - All TensorLists and ScalarLists must have the same number of elements.
// - Corresponding tensors must have the same size.

Its done in `check_foreach_api_restrictions`. 

Added a comment about it.